docs: RTD migration (#708)
canonical/opensearch-operator: docs/how-to/scale-horizontally.md

(how-to-scale-horizontally)=
# How to safely scale down

This page outlines the general steps to follow when scaling down Charmed OpenSearch
to prevent data loss and ensure the deployment remains highly available.

To see an example of scaling down a real deployment scenario, check the following page from the
Charmed OpenSearch Tutorial: [6. Scale horizontally](tutorial-6-scale-horizontally).

```{warning}
* The following steps are for removing one single Juju unit (node). This may be repeated as many times as necessary, but **never remove multiple units in the same command.** 
* In highly available deployments, **it is not safe to scale below 3 nodes**. 
```

## 1. Check cluster health before scaling down

> First of all, make sure that removing nodes is a safe operation to do.

Reviewer: Todo
```suggestion
First, ensure that removing nodes is a safe operation by checking the health of the cluster.
```

```shell
curl --cacert cert.pem -k -XGET https://10.180.162.96:9200/_cat/shards -u admin:admin_password
```

A general good course of action here would be to scale up (add a unit) to have a `green` state where all primary and replica shards are well assigned.

> To investigate why is your cluster in a `yellow` state. You can make the following call to have an explanation:

Reviewer: Todo
```suggestion
To investigate why your cluster is in a `yellow` state, run the following command:
```

```shell
curl --cacert cert.pem -k -XGET "https://10.180.162.96:9200/_cluster/allocation/explain?filter_path=index,shard,primary,**.node_name,**.node_decision,**.decider,**.decision,**.*explanation,**.unassigned_info,**.*delay"  -u admin:admin_password
```

<!-- What can we expect as an output?-->

> Depending on the output, there may be a different course of action. For example: scaling up, adding more storage to the existing nodes, or perhaps [manually re-route](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/cluster-reroute.html) the relevant shard manually to another node.

Reviewer: Todo
```suggestion
Depending on the output, there may be a different course of action. For example: scaling up, adding more storage to the existing nodes, or [manually re-routing](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/cluster-reroute.html) the relevant shard to another node.
```

:red_circle: Scaling down **is definitely not safe** to do, as some primary shards are not assigned. This is roughly equivalent to a `blocked` juju status.

The course of action to follow here is to add units to the cluster. To scale up by one unit, run the following command:

```shell
juju add-unit -n 1
```

```{note}
If the health color is `red` after removing a unit, the charm will attempt to block the removal of the node, giving the administrator the opportunity to scale up (add units).
```

<!-- TODO: clarify
**Note:** You'll notice we did not use the certificates to authenticate the curl requests above, in a real world example you should always make sure you verify your requests with the TLS certificates received from the `get-password` action.
i.e:

```
curl --cacert cert.pem -XGET https://admin:admin_password@10.180.162.96:9200/_cluster/health
``` 
-->

## 2. Scale down one unit

> Once you made sure that removing a unit is safe to do, you can proceed to removing **a single unit**. It is unsafe to remove more than one unit at a time.

Reviewer: Todo
```suggestion
Once you have verified that it is safe to remove a unit, proceed with removing **a single unit**. It is unsafe to remove more than one unit at a time.
```

## 3. Repeat cluster health check

> After removing **one unit**, depending on the roles of the said unit,

Reviewer: Todo
```suggestion
After removing **one unit**,
```

> the charm may reconfigure and restart a unit to balance the node roles.

Reviewer: Todo (is this still valid?)
```suggestion
the charm may reconfigure and restart another unit to balance the node roles.
```

You can monitor this with `juju status --watch 1s`.
<!-- what happens to each role?-->

> Make sure you wait for the whole application to stabilize before you consider removing further units.

Reviewer: Todo
```suggestion
Wait for the application to fully stabilize before removing additional units.
```
